### DISCUSSION TITLE
Pocketbase 400 errors

### DISCUSSION THREAD

**Discussion by bmingles** [pocketbase/pocketbase#5158] [Q&A]
My Pocketbase server keeps getting into a state where all requests result in `400` status errors:

I am using the JS SDK in browser. I see a call to `http://127.0.0.1:3002/api/collections/users/auth-with-password?expand=subs.org` in the network tab with an error in the console:
```javascript
ClientResponseError 400: ClientResponseError
    at pocketbase.js?v=f06b7a62:890:13
```

I am running my server with the `--dev` flag, but I don't see any error output. I only see:
```
SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '_pb_users_auth_' OR LOWER([[name]])='_pb_users_auth_' LIMIT 1
```

If I attempt to login to the admin UI, I also get a `400` status:
```
{"code":400,"message":"Something went wrong while processing your request.","data":{}}
```

I also don't see any error output, only:
```
SELECT count(*) FROM `_admins`
```

If I restart my server, everything starts working again.

This is happening frequently for me, and once it gets into this state, a server restart seems to be required. Is there any additional way to debug?

**Comment by ganigeorgiev** [comment]
_This seems related to your other question in https://github.com/pocketbase/pocketbase/discussions/5121._

It doesn't make sense a 400 error (or any other API response error) to "block" your server and most likely the issue you are experiencing is either caused by a custom hooks code (eg. deadlock due to using `app.Dao()` write in a transaction block) or some other deployment configuration (eg. issues with docker, network mounted disks, strange OS/filesystem virtualization conflict, etc.).
In general if it is the latter I'd expect at least a log entry in the stderr so if you have a systemd service you could check its log files.

In any case without a **minimal** reproducible example or at least some steps to simulate your environment, I don't think anyone will be able to assist you.

**Reply by bmingles** [reply]
@ganigeorgiev yes, this is related to my other question. I had just separated it since my earlier question was just about enabling debugging, and you had clearly answered that one.

The only thing I have in my `pb_hooks` directory are 2 util modules I use in my migrations. I wouldn't expect these to have any impact after migrations have run, and they don't have any `app.Dao` references, but I don't have experience here.

I am running on an Intel MacBook Pro via Docker compose (not sure if systemd applies here)

My Docker file is:
```Dockerfile
# Pocketbase server
FROM golang:1.22.4

WORKDIR /pb

COPY go.mod go.sum ./
RUN go mod download

COPY *.go ./

# Build
RUN CGO_ENABLED=0 GOOS=linux go build -o /pb/kmt

COPY ./pb_migrations /pb/pb_migrations

COPY ./pb_hooks /pb/pb_hooks

EXPOSE 8080

ENTRYPOINT [ "/pb/kmt", "serve", "--http=0.0.0.0:8080" ]
```

I am running this via Docker compose:
```yaml
  pb-kmt:
    image: pb-kmt
    build:
      context: pb-kmt
    # TODO: prod config should probably not include --dev flag
    command: --dev
    ports:
      - 3002:8080
    restart: always
    env_file:
      - ./.env.local
    volumes:
      - ./pb-kmt/pb_data:/pb/pb_data
      - ./pb-kmt/pb_migrations:/pb/pb_migrations
```
      
My `main.go` looks like:

```go
package main

import (
	"log"
	"os"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/plugins/jsvm"
	"github.com/pocketbase/pocketbase/plugins/migratecmd"
)

func main() {
    app := pocketbase.New()

    var hooksDir string
	app.RootCmd.PersistentFlags().StringVar(
		&hooksDir,
		"hooksDir",
		"",
		"the directory with the JS app hooks",
	)

	var hooksWatch bool
	app.RootCmd.PersistentFlags().BoolVar(
		&hooksWatch,
		"hooksWatch",
		true,
		"auto restart the app on pb_hooks file change",
	)

    var migrationsDir string
	app.RootCmd.PersistentFlags().StringVar(
		&migrationsDir,
		"migrationsDir",
		"",
		"the directory with the user defined migrations",
	)

    var automigrate bool
	app.RootCmd.PersistentFlags().BoolVar(
		&automigrate,
		"automigrate",
		true,
		"enable/disable auto migrations",
	)

    // load jsvm (hooks and migrations)
	jsvm.MustRegister(app, jsvm.Config{
		MigrationsDir: migrationsDir,
	})

    // migrate command (with js templates)
	migratecmd.MustRegister(app, app.RootCmd, migratecmd.Config{
		TemplateLang: migratecmd.TemplateLangJS,
		Automigrate:  automigrate,
		Dir:          migrationsDir,
	})

    // serves static files from the provided public dir (if exists)
    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        e.Router.GET("/", func(c echo.Context) error {
            return c.String(404, "Not Found")
        })

        e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./pb_public"), false))

        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Reply by ganigeorgiev** [reply]
The above code is the same as in your other question and it is not enough to reproduce your problem.

I've just checked the stderr and activity logs of 2 servers (the live demo and presentator.io) and there is nothing unusual. Other than the failed bot 404 and 405 requests, based on the activity logs for the past 7 days of presentator.io there are ~100 400 failed requests and the application hasn't been restarted since its 3.2.5 release ~2-3 weeks ago and everything seems to work correctly.

I'm not ruling out a PocketBase issue, but again without some clear reproducible steps I'm not willing to invest time on this as there is nothing much I can help with.

**Reply by bmingles** [reply]
Sadly, I don't have a way to reproduce it on demand. It just seems to get in this state after I have worked with my app for a bit. I'll post more info if I can find a way to reliably reproduce it. Was just hoping there were some logs somewhere I was missing or something I could dig further into. Thanks.

**Reply by ganigeorgiev** [reply]
Based on the [`BadRequestError` usage](https://github.com/search?q=repo%3Apocketbase%2Fpocketbase%20BadRequestError&type=code) there should be either a detailed public message or extra `err` data that will be printed with `--dev`.

You've mentioned that you have `pb_hooks` code that is not actually used but I have no way to confirm this (_I have no way to even confirm the reports from your initial post are even correct or what version of PocketBase do you use_).

Either you are not testing/reporting it properly or there is an error that I've missed to provide context but I can't find it.

The only way at the moment that I could think of that could result in a generic 400 error without some details when `--dev` is enabled is:

- you have custom hooks code that throws/return blank error
- there is http cache enabled somewhere (eg. on reverse proxy level or if you use Next.js) that returns a cached response from earlier attempts without the `--dev` flag (but even then there should be at least a log in the Admin UI > Logs section if the request has reached the server)

**Reply by bmingles** [reply]
Ok, I've pushed the Pocketbase pieces to a dedicated repo (all of the code is now included):
https://github.com/bmingles/pb-spike

To clarify the `pb_hooks` piece, the 2 modules it contains do get run, they just only get used by migrations. I wouldn't expect that to cause issues after the server has successfully run migrations and has been running a while, but this is an assumption on my part which may not be true.

https://github.com/bmingles/pb-spike/tree/main/pb_hooks

This also includes my `go.mod` which shows `github.com/pocketbase/pocketbase v0.22.13`

**Reply by ganigeorgiev** [reply]
I've cloned the repo but I wasn't able to reproduce the issue.

I'm testing on Linux (Fedora 40), Docker 26.1.3 and docker-compose 2.27.1.

I've tried to spam the server with mixed valid and invalid `auth-with-password` and crud data requests (both create and update) but it still operates correctly.

It doesn't make sense the server to suddenly get into a state where it will return only 400 errors.

There is definitely something more to your environment and/or deployment setup that I'm not able to simulate, especially since that you are using Docker on Mac, but until there are clear reproducible steps there is nothing much I can help further with.

**Reply by bmingles** [reply]
Thank you for investigating. I appreciate it. I will post reproduction steps if / when I can identify them. I may try to run on a different machine to rule out anything I’m doing vs this likely being something specific to my Mac

**Reply by bmingles** [reply]
Not sure if this is related, but I did notice this error in my logs today:
```
checkin-pb-kmt-1  | [82.00ms] SELECT DISTINCT `person`.* FROM `person` LEFT JOIN `users` `__auth_users` ON `__auth_users`.`id`='lcn7bil8mriwd1c' LEFT JOIN json_each(CASE WHEN json_valid([[__auth_users.subs]]) THEN [[__auth_users.subs]] ELSE json_array([[__auth_users.subs]]) END) `__auth_users_subs_je` LEFT JOIN `subscription` `__auth_users_subs` ON [[__auth_users_subs.id]] = [[__auth_users_subs_je.value]] LEFT JOIN `organization` `__auth_users_subs_org` ON [[__auth_users_subs_org.id]] = [[__auth_users_subs.org]] WHERE (`person`.`id` IN ('bgq5zoy2t8971ww', 'ga0035heiui4ukl', 'xxirqg6k1uy7j36', '1zwgsfgy5au1o0e', '90v5d77kjeaffr6', 'h0329yyz2m2u87g')) AND (COALESCE([[__auth_users_subs_org.id]], '') = COALESCE([[person.org]], ''))
checkin-pb-kmt-1  | DEBUG Failed to enrich list records
checkin-pb-kmt-1  | └─ map[error:Failed to expand: map[members:unable to open database file (14); failed query: SELECT DISTINCT `person`.* FROM `person` LEFT JOIN `users` `__auth_users` ON `__auth_users`.`id`={:p0} LEFT JOIN json_each(CASE WHEN json_valid([[__auth_users.subs]]) THEN [[__auth_users.subs]] ELSE json_array([[__auth_users.subs]]) END) `__auth_users_subs_je` LEFT JOIN `subscription` `__auth_users_subs` ON [[__auth_users_subs.id]] = [[__auth_users_subs_je.value]] LEFT JOIN `organization` `__auth_users_subs_org` ON [[__auth_users_subs_org.id]] = [[__auth_users_subs.org]] WHERE (`person`.`id` IN ({:p1}, {:p2}, {:p3}, {:p4}, {:p5}, {:p6})) AND (COALESCE([[__auth_users_subs_org.id]], '') = COALESCE([[person.org]], ''))]]
```

**Reply by ganigeorgiev** [reply]
The enrich error itself is not critical, but the `unable to open database file (14)` is.

This could be returned for a lot of reasons - issues with the mounted volume, permission errors, lack of disk space to create the temporary side-car wal files, etc.

It is most likely something related to the Docker-Mac setup and its configuration/permissions.

**Reply by ganigeorgiev** [reply]
I'm not sure if this apply for Mac, but it is also possible that you've hit the max allowed file descriptors for your env (see https://pocketbase.io/docs/going-to-production/#increase-the-open-file-descriptors-limit; _the Docker container usually should inherit the ulimit settings from the host_).

**Reply by ganigeorgiev** [reply]
I forgot to mention that the above error could be also returned if you have an external process that interacts with the `pb_data` directory or operates directly with the `pb_data/data.db` file (Litestream, Marmot, cloud backup software, antivirus or anything else that will lock the db file and prevent the PocketBase process from reading it).

But in any case I don't think the issue is something in PocketBase. If you are not able to identify the culprit, try to run PocketBase directly on your Mac without Docker to rule out external tooling setup/permissions misconfiguration.

**Reply by bmingles** [reply]
I'm seeing some evidence that the culprit may have been my particular usage of `pocketbase-typegen` to generate types from my database. Namely, I was using the direct connect option to the sql db. I've reconfigured it to run against the REST apis instead. Still needs more testing, but so far haven't seen the flakiness since the change.
